apps/prairielearn/src/pages/instructorInstanceAdminSettings/components/SelfEnrollmentSettings.tsx

```diff
@@ -86,6 +152,158 @@ export function SelfEnrollmentSettings({
+        </div>
+      </div>
+
+      <div class="mb-3 form-check">
+        <input
+          class="form-check-input"
+          type="checkbox"
+          id="self_enrollment_requires_secret_link"
+          disabled={!canEdit || !selfEnrollmentEnabled || !enrollmentManagementEnabled}
+          {...register('self_enrollment_requires_secret_link')}
+          name="self_enrollment_requires_secret_link"
+        />
+        <label class="form-check-label" for="self_enrollment_requires_secret_link">
+          Self-enrollment requires secret link
+        </label>
+        <div class="small text-muted">
+          If enabled, self-enrollment requires a secret link to enroll. If disabled, any link to the
+          course instance will allow self-enrollment.
+        </div>
+      </div>
+
+      <div class="mb-3 form-check">
+        <input
+          class="form-check-input"
+          type="checkbox"
+          id="disable_self_enrollment_after_date"
+          disabled={!canEdit || !selfEnrollmentEnabled || !enrollmentManagementEnabled}
+          checked={showDateInput}
+          onChange={(e) => {
+            const target = e.target as HTMLInputElement;
+            setShowDateInput(target.checked);
+
+            // Clear the date when unchecking
+            if (!target.checked) {
+              setValue('self_enrollment_enabled_before_date', '');
+            }
+          }}
+        />
+        <label class="form-check-label" for="disable_self_enrollment_after_date">
+          Disable self-enrollment after specified date
+        </label>
+        <div class="small text-muted">
+          If enabled, self-enrollment will be disabled after the specified date.
+        </div>
+        {showDateInput && (
+          <>
+            <input
+              type="datetime-local"
+              aria-label="Self-enrollment enabled before date"
+              class={clsx(
+                'form-control mt-2',
+                errors.self_enrollment_enabled_before_date && 'is-invalid',
+              )}
+              disabled={!canEdit || !selfEnrollmentEnabled || !enrollmentManagementEnabled}
+              {...register('self_enrollment_enabled_before_date', {
+                required: showDateInput ? 'Date is required' : false,
+              })}
+              // If no date is set, don't include the name in the form data.
+              {...(selfEnrollmentEnabledBeforeDate.length > 0
+                ? { name: 'self_enrollment_enabled_before_date' }
+                : {})}
+            />
+            {errors.self_enrollment_enabled_before_date && (
+              <div class="invalid-feedback">
+                {errors.self_enrollment_enabled_before_date.message}
+              </div>
+            )}
+            <small class="form-text text-muted">
+              After this date, self-enrollment will be disabled.
+            </small>
+          </>
+        )}
+      </div>
+
+      {selfEnrollmentEnabled && (
```

Comment:
To avoid confusion, based on the enrollment settings, a link is either shown, not shown, or the self-enrollment link is shown.